### PR TITLE
Add Service Annotation Support

### PR DESCRIPTION
**Pull request by gpachecoNI**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-python/blob/main/CONTRIBUTING.md).

TODO: Check the above box with an 'x' indicating you've read and followed [CONTRIBUTING.md](https://github.com/ni/measurement-services-python/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?

- Add service Annotations support to the Measurement Services. The user can add any annotations but our examples will have the following annotations: description, collection, and tags.
- Adds the new proto for the discovery_service

### Why should this Pull Request be merged?

[AB#2408662](https://dev.azure.com/ni/DevCentral/_workitems/edit/2408662/)

### What testing has been done?

- Update unit test and include examples with the common annotations and user random annotations. Testing Integer, arrays, and dictionaries being added as an annotation. 
- Add cases to the create_measurment_service to ensure we cover all cases. 

### Open Issues
- The generator won't match with these changes, but I am currently working on updating it.

### CODE REVIEW

**Comment by gpachecoNI** [comment]
> The discovery_service.proto changes aren't expected to go into ASW until the release branch is made. So we should wait to check these in until Curtis' changes go in, right? https://dev.azure.com/ni/DevCentral/_git/ASW/pullrequest/507369

Yes, I am not planning to submit.

**Comment by bkeryan** [comment]
https://github.com/ni/measurementlink-python/pull/292 changes the codegen script. You will need to rerun it.

File: examples/sample_measurement/SampleMeasurement.serviceconfig
Comment (cumitche) [inline examples/sample_measurement/SampleMeasurement.serviceconfig:9]: Change these annotations to be about the actual sample measurement service.  The description should describe what the example actually does.  The collection should be "NI.Examples".  The tags should just be [].

Comment (dixonjoel) [inline examples/sample_measurement/SampleMeasurement.serviceconfig:9]: We should update the template files to include these standard annotations and suggest how the user should fill them in. We should also use the existing variables like 'display name' to fill them in when generated or add them as optional command-line parameters.

Comment (dixonjoel) [reply examples/sample_measurement/SampleMeasurement.serviceconfig:9]: This can and probably should be done in a separate PR.

Comment (gpachecoNI) [reply examples/sample_measurement/SampleMeasurement.serviceconfig:9]: Yeah, I noticed this once I looked at the example US.  Will fix.

Comment (gpachecoNI) [reply examples/sample_measurement/SampleMeasurement.serviceconfig:9]: Will revert this change since I made the correct fix in my PR of updating examples.

File: tests/assets/test_pb2.pyi
Comment (gpachecoNI) [inline tests/assets/test_pb2.pyi:78]: @WesleyTangNationalInstruments  @bkeryan 
 I added these 2 lines here since when I ran the updated version of mypy tests I was getting this error 
tests\unit\test_serializer.py:337: error: "MeasurementParameter" has no attribute "enum_data"  [attr-defined]
tests\unit\test_serializer.py:338: error: "MeasurementParameter" has no attribute "enum_array_data"; maybe "int32_array_data"

Comment (gpachecoNI) [reply tests/assets/test_pb2.pyi:78]: It was an update that did Brad at the same time you submitted Wes. 
https://github.com/ni/measurementlink-python/pull/292

**Comment by gpachecoNI** [comment]
I am adding this comment to ensure we don't submit by mistake.

File: ni_measurementlink_service/_internal/discovery_client.py
Comment (gpachecoNI) [inline ni_measurementlink_service/_internal/discovery_client.py:107]: I am adding this comment to ensure I don't submit it by mistake.

**Comment by bkeryan** [comment]
FYI, I created a releases/1.1 branch and updated main to version 1.2.0-dev0 in order to unblock these PRs.

File: ni_measurementlink_service/_internal/discovery_client.py
Comment (cumitche) [inline ni_measurementlink_service/_internal/discovery_client.py:115]: I don't think this will handle the possibility of complex types that the user could provide.  A dictionary that contains a dictionary that contains an array of strings will need to be converted to embedded JSON.  Simply joining with a space won't do that, and embedded JSON usually removes all whitespace outside of a string.

File: tests/unit/test_discovery_client.py
Comment (cumitche) [inline tests/unit/test_discovery_client.py:19]: The annotations map in the request object is string to string.  All of the values needs to be changes to embedded JSON strings.

File: ni_measurementlink_service/_internal/stubs/ni/measurementlink/measurement/v1/measurement_service_pb2.pyi
Comment (bkeryan) [inline ni_measurementlink_service/_internal/stubs/ni/measurementlink/measurement/v1/measurement_service_pb2.pyi:48]: Did you run Black on the codegen? I thought we stopped doing that. Please update the excludes in pyproject.toml so this doesn't happen.

File: ni_measurementlink_service/measurement/info.py
Comment (bkeryan) [inline ni_measurementlink_service/measurement/info.py:46]: Python doesn't use C++/C# syntax.

```suggestion
        annotations (Dict[str,str]): Dict that contains extra information of the measurement.
```

File: ni_measurementlink_service/measurement/service.py
Comment (bkeryan) [inline ni_measurementlink_service/measurement/service.py:189]: Simplify using get(), items(), and dict comprehensions.

The loop variable should be singular.

You probably don't need a type hint here because `json.dumps` returns a `str`: https://github.com/python/typeshed/blob/main/stdlib/json/__init__.pyi#L13

```suggestion
        service_annotations_string = {
            key: json.dumps(value, separators=(",", ":")
            for key, value in service.get("annotations", {}).items()
        }
```

File: tests/assets/example.v3.serviceconfig
Comment (bkeryan) [inline tests/assets/example.v3.serviceconfig:1]: What does v3 in the filename mean?

File: tests/unit/test_discovery_client.py
Comment (bkeryan) [inline tests/unit/test_discovery_client.py:19]: Provided annotations for what? You have a bunch of different test serviceconfigs and this seems to have merged values from all of them. What happens when they confict?

FYI, the "provided" in `_PROVIDED_MEASUREMENT_SERVICES` refers to the `providedInterfaces` key and may not be relevant here.

File: tests/unit/test_discovery_client.py
Comment (bkeryan) [inline tests/unit/test_discovery_client.py:101]: Why change this to strict equality? >= means superset.

File: tests/unit/test_discovery_client.py
Comment (bkeryan) [inline tests/unit/test_discovery_client.py:105]: You can compare dicts with == without looping over them.

File: tests/unit/test_service.py
Comment (bkeryan) [inline tests/unit/test_service.py:331]: Iterating over an empty dict is benign.

File: tests/unit/test_service.py
Comment (bkeryan) [inline tests/unit/test_service.py:332]: Again, you can compare dicts with == without looping.

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: ni_measurementlink_service/_internal/stubs/ni/measurementlink/measurement/v1/measurement_service_pb2.pyi
Comment (gpachecoNI) [reply ni_measurementlink_service/_internal/stubs/ni/measurementlink/measurement/v1/measurement_service_pb2.pyi:48]: I did run it. Will take a look at the pyproject.toml

File: ni_measurementlink_service/_internal/discovery_client.py
Comment (gpachecoNI) [reply ni_measurementlink_service/_internal/discovery_client.py:112]: Nice, it does work

File: ni_measurementlink_service/_internal/stubs/ni/measurementlink/measurement/v1/measurement_service_pb2.pyi
Comment (gpachecoNI) [reply ni_measurementlink_service/_internal/stubs/ni/measurementlink/measurement/v1/measurement_service_pb2.pyi:48]: I don't think is there. Do you want me to add this file? Here? 
extend-exclude ='((.+_pb2_grpc)|(.+_pb2))+(.py)$' # Regex to files with _pb2_grpc.py and _pb2.py Auto Generated Code are excluded

File: tests/unit/test_discovery_client.py
Comment (gpachecoNI) [reply tests/unit/test_discovery_client.py:19]: _PROVIDED_ANNOTATIONS is similar to _PROVIDED_MEASUREMENT_SERVICES.. the test just use this to form the ServiceInfo and then checks with itself.   Both are used for the test only.  

The reason I have like a mix, is that I wanted to add more cases to this single test.

File: tests/assets/sample_measurement_test_pb2.pyi
Comment (bkeryan) [inline tests/assets/sample_measurement_test_pb2.pyi:27]: These .pyi changes look like reformatting with Black. Revert them.

File: tests/unit/test_discovery_client.py
Comment (bkeryan) [reply tests/unit/test_discovery_client.py:19]: I still think it's confusing to combine annotations for 6 versions of the same service into a single dictionary. Does the feature really work this way? You can add multiple .serviceconfig files for the same service and the discovery service merges their annotations?

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: tests/assets/example.v3.serviceconfig
Comment (bkeryan) [reply tests/assets/example.v3.serviceconfig:1]: Also, what happened to v2?

File: tests/unit/test_discovery_client.py
Comment (gpachecoNI) [reply tests/unit/test_discovery_client.py:19]: On this test we build the service_info, so there is no different versions of .serviceconfig  
The Annotations I added all could be part of a .serviceconfig. And is not supposed to look good, is supposed to be the json embedded representation.

Comment (gpachecoNI) [reply tests/unit/test_discovery_client.py:19]: I guess I am not understanding the confusion here.  It is being used exactly as _PROVIDED_MEASUREMENT_SERVICES.. I know that refers to providedInterfaces. And my Annotations refers to the same..  it is a valid case.

Comment (bkeryan) [reply tests/unit/test_discovery_client.py:19]: Okay, now I see. These strings match multiple `.serviceconfig` files, but that is a coincidence because you copy/pasted them. This test doesn't use the `.serviceconfig` files at all.

Please change these strings so they don't match the `.serviceconfig` files. Otherwise, someone making changes to one might assume both copies are related and have to be kept in sync.

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: tests/unit/test_discovery_client.py
Comment (gpachecoNI) [reply tests/unit/test_discovery_client.py:19]: Yeap, that I can do.

File: tests/assets/example.v3.serviceconfig
Comment (gpachecoNI) [reply tests/assets/example.v3.serviceconfig:1]: Ahhh I was using it just as Example 3...

**Review by bkeryan — APPROVED** [review]

**Comment by bkeryan** [comment]
Both Joel and I have approved this, so please merge it. If you don't have access, let us know.

**Pull request merged**
